plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/Pom.java

```diff
@@ -270,6 +280,66 @@ void updateDependencies(Collection<Library> selectedLibraries,
     }   
   }
 
+  private void createBOMIfNeeded(XPath xpath) throws CoreException {
+    try {
+      Element bomElement = (Element) xpath.evaluate(
+          "//m:dependencyManagement/m:dependencies/m:dependency[m:groupId='"
+              + Bom.GOOGLE_BOM_GROUP_ID + "'][m:artifactId='" + Bom.GOOGLE_BOM_ARTIFACT_ID + "']",
+          document.getDocumentElement(), XPathConstants.NODE);
+      if (bomElement == null) {
+        Element dependencies = null;
+        NodeList dependenciesNodes =
+            (NodeList) xpath.evaluate("//m:dependencyManagement/m:dependencies",
+                document.getDocumentElement(), XPathConstants.NODESET);
+        if (dependenciesNodes.getLength() > 0) {
+          dependencies = (Element) dependenciesNodes.item(0);
+        } else {
+          dependencies = document.createElementNS(NAMESPACE_URI, DEPENDENCIES);
+          Node dependencyManagement = (Node) xpath.evaluate("//m:dependencyManagement",
+              document.getDocumentElement(), XPathConstants.NODE);
+          if (dependencyManagement == null) {
+            dependencyManagement = document.createElementNS(NAMESPACE_URI, DEPENDENCY_MANAGEMENT);
+          }
+          dependencyManagement.appendChild(dependencies);
+          document.getDocumentElement().appendChild(dependencyManagement);
+        }
+
+        if (dependencies == null)
+          return;
+
+        Element dependency = document.createElementNS(NAMESPACE_URI, DEPENDENCY);
+        dependency.appendChild(createChild(GROUP_ID, Bom.GOOGLE_BOM_GROUP_ID));
+        dependency.appendChild(createChild(ARTIFACT_ID, Bom.GOOGLE_BOM_ARTIFACT_ID));
+        dependency.appendChild(createChild(TYPE, Bom.GOOGLE_BOM_TYPE));
+        dependency.appendChild(createChild(SCOPE, Bom.GOOGLE_BOM_SCOPE));
+
+        String version = getBestVersion(Bom.GOOGLE_BOM_GROUP_ID, Bom.GOOGLE_BOM_ARTIFACT_ID);
+        if (version != null && !version.isEmpty()) {
```

Comment:
I'm not decisive yet, but I wonder if we should fail if we fail to retrieve `version`, because otherwise you end up adding a broken `<dependency>`.

Reply:
Modified method # getBestVersion()
new method returns "MavenCoordinates.LATEST_VERSION" if version is null. Someone please review it.
private String getBestVersion(String groupId, String artifactId) {
    ArtifactVersion latestVersion = ArtifactRetriever.DEFAULT.getBestVersion(groupId, artifactId);
    return latestVersion != null ? latestVersion.toString() : MavenCoordinates.LATEST_VERSION;
  }